spectrometer.py

```diff
@@ -1,303 +1,331 @@
+    return int(255 * color[0] * factor), int(255 * color[1] * factor), int(255 * color[2] * factor)
+
+
+def take_picture(camera, name, shutter):
+    camera.vflip = True
+    camera.framerate = Fraction(1, 6)
+    camera.shutter_speed = shutter
+    camera.iso = 100
+    camera.exposure_mode = 'off'
+    camera.awb_mode = 'off'
+    camera.awb_gains = (1, 1)
+
+    time.sleep(3)
+
+    raw_filename = name + "_raw.jpg"
+    camera.capture(raw_filename, resize=(1296, 972))
+    return raw_filename
+
+
+def draw_graph(draw, pic_pixels, aperture: object, spectrum_angle, wavelength_factor):
+    aperture_height = aperture['h'] / 2
+    step = 1
+    last_graph_y = 0
+    max_result = 0
+    results = OrderedDict()
+    for x in range(0, int(aperture['x'] * 7 / 8), step):
+        wavelength = (aperture['x'] - x) * wavelength_factor
+        if 1000 < wavelength or wavelength < 380:
+            continue
+
+        # general efficiency curve of 1000/mm grating
+        eff = (800 - (wavelength - 250)) / 800
+        if eff < 0.3:
+            eff = 0.3
+
+        # notch near yellow maybe caused by camera sensitivity
+        mid = 575
+        width = 10
+        if (mid - width) < wavelength < (mid + width):
+            d = (width - abs(wavelength - mid)) / width
+            eff = eff * (1 - d * 0.1)
+
+        # up notch near 590
+        mid = 588
+        width = 10
+        if (mid - width) < wavelength < (mid + width):
+            d = (width - abs(wavelength - mid)) / width
+            eff = eff * (1 + d * 0.1)
+
+        y0 = math.tan(spectrum_angle) * (aperture['x'] - x) + aperture['y']
+        amplitude = 0
+        ac = 0.0
+        for y in range(int(y0 - aperture_height), int(y0 + aperture_height), 1):
+            r, g, b = pic_pixels[x, y]
+            q = r + b + g * 2
+            if y < (y0 - aperture_height + 2) or y > (y0 + aperture_height - 3):
+                q = q * 0.5
+            amplitude = amplitude + q
+            ac = ac + 1.0
+        amplitude = amplitude / ac / eff
+        # amplitude=1/eff
+        results[str(wavelength)] = amplitude
+        if amplitude > max_result:
+            max_result = amplitude
+        graph_y = amplitude / 50 * aperture_height
+        draw.line((x - step, y0 + aperture_height - last_graph_y, x, y0 + aperture_height - graph_y), fill="#fff")
+        last_graph_y = graph_y
+    draw_ticks_and_frequencies(draw, aperture, spectrum_angle, wavelength_factor)
+    return results, max_result
+
+
+def draw_ticks_and_frequencies(draw, aperture, spectrum_angle, wavelength_factor):
+    aperture_height = aperture['h'] / 2
+    for wl in range(400, 1001, 50):
+        x = aperture['x'] - (wl / wavelength_factor)
+        y0 = math.tan(spectrum_angle) * (aperture['x'] - x) + aperture['y']
+        draw.line((x, y0 + aperture_height + 5, x, y0 + aperture_height - 5), fill="#fff")
+        font = ImageFont.truetype('/usr/share/fonts/truetype/lato/Lato-Regular.ttf', 12)
```

Comment:
Hard coding a filesystem path here would be a flag were this production code because it a) creates a hard external dependency and b) consequently makes the function difficult to test and/or reuse easily in other contexts.